JAMES-2372 Fixing multiple accounts with Fetchmail

MaximeBrunetPolyconseil opened the pull request:
This PR aims to fix https://issues.apache.org/jira/browse/JAMES-2372.

There was a bug with the increment of one variable but also with getting the configuration for each account.

I did a test but I'm not certain about the correct use and configuration of users with different types for accounts.

Arsnael commented on server/protocols/fetchmail/src/test/java/org/apache/james/fetchmail/FetchMailTest.java:
```diff
@@ -0,0 +1,68 @@
+        try (MockedConstruction<ParsedConfiguration> mockParsedConfiguration = Mockito.mockConstruction(ParsedConfiguration.class)) {
+            UsersRepository fieldLocalUsers = mock(UsersRepository.class);
+            when(fieldLocalUsers.countUsers()).thenReturn(4);
+            when(fieldLocalUsers.list()).thenReturn(List.of(Username.of("user23@domain.com")).iterator()).thenReturn(List.of(Username.of("user23@domain.com")).iterator());
```
```suggestion
            when(fieldLocalUsers.list()).thenReturn(List.of(Username.of("user23@domain.com")).iterator());
```
think is enough :)

chibenwa replied:
Could we indent this?

MaximeBrunetPolyconseil replied:
Thank you for the review. The thenReturn is added twice because it is called twice in the test. Otherwise the first iterator is returned on second call so it has already been iterated over and returns no value on hasNext.

I have indented the file using the formatting.xml mentioned https://james.apache.org/contribute.html. Am i missing something?

Arsnael replied:
What @chibenwa means by indentation is something looking like this:

```
when(fieldLocalUsers.list())
    .thenReturn(List.of(Username.of("user23@domain.com")).iterator())
    .thenReturn(List.of(Username.of("user23@domain.com")).iterator());
```

> The thenReturn is added twice because it is called twice in the test.

Ah maybe yes, those iterators can be tricky sometimes :) Fair enough!

Arsnael reviewed (commented):
Except a minor remark to fix, I think looks good enough to me. Thank you for this!

chibenwa commented on server/protocols/fetchmail/pom.xml:
```diff
@@ -94,6 +94,12 @@
         </dependency>
+
+        <dependency>
+            <groupId>org.mockito</groupId>
```
POM file needs alphabetic ordering. THis thus need to ge before SLF4J.

chibenwa reviewed (commented):
Hi. Thanks for the fix!

I am quite curious on why and how you encountered this bug: FetchMail is not a very used component of James, is not wired with James.

Did you suceeded to run it with Spring?

What do you use FetchMail for?

The pull request otherwize looks fine to me.

MaximeBrunetPolyconseil commented:
> Hi. Thanks for the fix!
> 
> I am quite curious on why and how you encountered this bug: FetchMail is not a very used component of James, is not wired with James.
> 
> Did you suceeded to run it with Spring?
> 
> What do you use FetchMail for?
> 
> The pull request otherwize looks fine to me.

I have not been using FetchMail myself, i wanted to contribute to the project and went for some tickets in Jira that were tagged as newbie. It seemed like a useful ticket to do i suppose.

MaximeBrunetPolyconseil commented:
> > Hi. Thanks for the fix!
> > I am quite curious on why and how you encountered this bug: FetchMail is not a very used component of James, is not wired with James.
> > Did you suceeded to run it with Spring?
> > What do you use FetchMail for?
> > The pull request otherwize looks fine to me.
> 
> I have not been using FetchMail myself, i wanted to contribute to the project and went for some tickets in Jira that were tagged as newbie. It seemed like a useful ticket to do i suppose.
> 
> I'm going to try and run it with Spring to make sure it's working correctly.

Hello, I ran the Spring application with FetchMail enabled with two accounts added and they were both added with the right values and the fieldSequence incremented. So it is working.

![image](https://github.com/apache/james-project/assets/148060230/be9e8afb-2e92-4e23-b399-6b736e5cb9f8)

Though I'm not sure why some unit tests are failing. They have nothing to do with this MR (other modules) and they also fail on master for me.

chibenwa commented on server/protocols/fetchmail/src/test/java/org/apache/james/fetchmail/FetchMailTest.java:
```diff
@@ -0,0 +1,76 @@
+    @Test
+    void shouldAddAccountWithGoodDataFetchMail() throws Exception {
+        try (MockedConstruction<ParsedConfiguration> mockParsedConfiguration = Mockito.mockConstruction(ParsedConfiguration.class)) {
+            UsersRepository fieldLocalUsers = mock(UsersRepository.class);
+            when(fieldLocalUsers.countUsers()).thenReturn(4);
+            when(fieldLocalUsers.list())
+                    .thenReturn(List.of(Username.of("user23@domain.com")).iterator())
+                    .thenReturn(List.of(Username.of("user23@domain.com")).iterator());
```
Maybe use `MemoryUserRepository` in order to avoid mocked data?

This way we would avoid using a mock and all related errors.

MaximeBrunetPolyconseil replied:
Thank you for your review. I removed the mock for `UserRepository` and used `MemoryUserRepository`. I also improved the test by adding only one alllocal account as it is only possible to have one for a given fetchmail configuration.

They are no errors otherwise in the new tests. The errors are coming from other modules, the detail is in the CI run. But it will be fix eventually.

chibenwa reviewed (commented):
Which error do you encounter running the unit tests?

Arsnael commented:
> Though I'm not sure why some unit tests are failing. They have nothing to do with this MR (other modules) and they also fail on master for me.

master branch is broken at the moment and we work on fixing it. Don't worry too much for the moment, we will rerun your PR when things are fixed :) and sorry for the temporary inconvenience

chibenwa reviewed (commented):
As we no longer use mockito we can cleanup the imports / the dependencies?

MaximeBrunetPolyconseil commented:
> As we no longer use mockito we can cleanup the imports / the dependencies?

Yes my bad. As I removed auto indentation, I also removed the cleanup of imports. I still use mockito to mock the construction of `ParsedConfiguration` since it would need way too much setup for no reason in the test.

Arsnael commented:
@chibenwa green! are you ok with latest changes now? Can we merge this?

chibenwa reviewed (approved):
Yes ;-)

Arsnael commented:
Merged! Thank you for your kind contribution @MaximeBrunetPolyconseil  :)
